### PR TITLE
 Fix GCM data block send. Fixes #297

### LINKED ISSUE #297: GCM rejecting data as "non-json"

### PR DESCRIPTION
**Pull request by jrconlin**
* pass "router_token" for senderid lookup
* Clean up comments to be more clear about what each token is.
* Add ability for clients to request "dry run"
* better exam of gcm data response block.
* fix test SenderID hash to match reality.

@bbangert r?

### CODE REVIEW

File: autopush/router/gcm.py [hunk 1]
```diff
@@ -43,12 +43,14 @@ def amend_msg(self, msg):
         msg["senderid"] = self.creds.get("senderID")
         return msg
 
-    def register(self, uaid, router_data):
-        """Validate that a token is in the ``router_data``"""
-        if not router_data.get("token"):
-            raise self._error("connect info missing 'token'", status=401)
+    def register(self, uaid, router_data, router_token=None, *kwargs):
+        """ Validate that the GCM Instance Token is in the ``router_data``"""
```
Comment (bbangert) [inline autopush/router/gcm.py]: Nit: leading space seems to have snuck in here.

File: autopush/router/gcm.py [hunk 2]
```diff
@@ -43,12 +43,14 @@ def amend_msg(self, msg):
...
+        if "token" not in router_data:
+            raise self._error("connect info missing GCM Instance 'token'",
+                              status=401)
         # Assign a senderid
-        router_data["creds"] = self.creds = self.senderIDs.choose_ID()
+        router_data["creds"] = self.creds = \
```
Comment (bbangert) [inline autopush/router/gcm.py]: This didn't fit on one line?

Comment (bbangert) [inline autopush/router/gcm.py]: ``self.creds``, which is on a single router instance used concurrently from multiple paths keeps changing its state based on the router_token of the last register call?

File: autopush/router/gcm.py [hunk 3]
```diff
@@ -98,12 +99,14 @@ def _route(self, notification, router_data):
             data['con'] = con
             data['enc'] = enc
 
+        # registration_ids are the GCM instance tokens (specified during
+        # registration.
         payload = gcmclient.JSONMessage(
-            registration_ids=[router_data["token"]],
+            registration_ids=[router_data.get("token")],
```
Comment (bbangert) [inline autopush/router/gcm.py]: If there's no token, this will be a list of [None]. Will that cause any issues?

File: autopush/router/gcm.py [hunk 4]
```diff
@@ -98,12 +99,14 @@ def _route(self, notification, router_data):
...
             collapse_key=self.collapseKey,
             time_to_live=self.ttl,
-            dry_run=self.dryRun,
-            data=udata,
+            dry_run=self.dryRun or "dryrun" in router_data,
```
Comment (bbangert) [inline autopush/router/gcm.py]: Remembering the order of operation precedence isn't my strong-point so I usually () clauses like this just to be sure. Nick actually found an error in older GCM code due to operator precedence working in an unintentional way.

File: autopush/senderids.py [hunk 5]
```diff
@@ -129,11 +129,11 @@ def senderIDs(self):
         """Return a list of senderIDs"""
         return self._senderIDs
 
-    def get_ID(self, id):
+    def get_ID(self, id=None):
         """Return the associated record for a given SenderID"""
+        if id is None:
```
Comment (bbangert) [inline autopush/senderids.py]: ``if not id:`` ? Or does this need to specifically check that its ``None``?

File: autopush/router/gcm.py (hunk 3 above)
Comment (jrconlin) [inline autopush/router/gcm.py]: It will be converted to JSON [null] which will be handled by the GCM call.

File: autopush/senderids.py (hunk 5 above)
Comment (jrconlin) [inline autopush/senderids.py]: I was originally thinking that we might reserve "" as a default defined in the senderid load table, but that might be overthinking and it's an uncommon use case at best.

**Comment by bbangert** [comment]
r+ pending CI sign-off

**Pull request merged**